src/OpenTelemetry.AutoInstrumentation/Configurations/ResourceConfigurator.cs

```diff
@@ -26,6 +26,8 @@ public static void Configure(ResourceBuilder resourceBuilder)
 
         resourceBuilder
             .AddTelemetrySdk()
+            .AddEnvironmentVariableDetector()
+            .AddServiceNameDetector()
```

Comment:
I don't like adding the environment variable detector "again" since it is part of the default resource builder created by the SDK. What we actually want is to set "service.name" if it is starting with "unknown_service:" (the bad default). So instead of checking if it was defined via env var, it makes more sense to build the resources and check if we should use our fallback or not. This won't look neat because of the `ResourceBuilder` interface but it seems to be what we should do. Something like:

```csharp
        resourceBuilder
            .AddTelemetrySdk()
            .AddAttributes(new KeyValuePair<string, object>[] { new(Constants.Tracer.AutoInstrumentationVersionName, Constants.Tracer.Version) });

        var resource = resourceBuilder.Build();
        var serviceNameValue = resource.Attributes.Where(kvp => kvp.Key == "service.name").Select(kvp => (kvp.Value as string)).FirstOrDefault();
        if (serviceNameValue is not null && serviceNameValue.StartsWith("unknown_service:"))
        {
            // Add fallback service name
            resourceBuilder.AddAttributes(new KeyValuePair<string, object>[] { new("service.name", FallbackServiceName()) });
        }
```

In the end, `Build()` is going to be called twice, one by us and one by the SDK for each signal. However, this removes the risk of "service.name" being explicitly set and automatic instrumentation overwriting it. An optimization with a very minimal effect that we could do is to cache the fallback service name so different signals don't need to recompute it - although I think we can go without it.

Reply:
@pjanotti, APi should allow us to use empty resource, something like:
ResourceBuilder.CreateEmpty().AddMethodsUsedInSdkWithoutServiceNAmeDetector().AddOurDetector().AddServiceNAmeDetector.

Reply:
checking for `unknown_service` is probably safer also, since there are also 2 ways to set service name. `OTEL_SERVICE_NAME` and `OTEL_RESOURCE_ATTRIBUTES`

Reply:
This is definitely interesting a bit closer to what I had originally thought about before going down the detector route. I'm all for simplicity, if we want to do this I'll create a new branch with less changes and a clean PR if that makes sense?

Reply:
@Kielek does the API allows one to check the keys that were already added to the resource? If I understood the SDK correctly, the DefaultResource is going to set the "unknown_service:<process_name>", then later the SDK uses another detector that uses an IConfiguration object and sets it again. What we actually want to do is check the actual value after all the "standard" detectors. If it ended up with the "unknown_service:<process_name>" we want to overwrite it.

I'm not sure if this is part of the spec or an API decision by the SDK team but it seems a reasonable case: have a way to plug a fallback for certain resource keys.

Reply:
@pjanotti `ResourceDetector.Detect()` is called on `ResourceBuilder.Build()`, so there is no knowledge before.

There seems to be a way to have 2 resource builders. We can build the second one when original is already built and then use `Resource.Merge(Resource other)` if we need to overwrite default results.

Reply:
@RassK can you make a concrete snippet of how you suggest using `Merge`?

Reply:
@pjanotti sure >

I think this:
* Removes need that ServiceNameDetector must check wheter service name was provided somehow.
* Does not build twice, specially the main ResourceBuilder, which can be quite costly.
* Could be moved to extension method like `_traceProvider!.GetResource().TryApplyAutoDefaults()`


```csharp
// this is Instrumentation.cs ~L143
_tracerProvider = builder.Build();

var resource = _tracerProvider!.GetResource();
var currentServiceName = resource.Attributes.FirstOrDefault(x => x.Key == "service.name").Value?.ToString();
if (currentServiceName == null || currentServiceName.StartsWith("unknown_service:"))
{
    var updatedResource = ResourceBuilder.CreateEmpty()
        .AddDetector(new AutoServiceNameDetector())
        .Build();

    resource.Merge(updatedResource);
};

Logger.Information("OpenTelemetry tracer initialized.");
```

Reply:
Thanks @RassK - I think we can achieve what we want without reflection combining it with @Kielek's suggestion. The key is that instead of using `ConfigureResource`, we should call `SetResourceBuilder` (like done for logging). So something like:

```csharp
    public static ResourceBuilder CreateResourceBuilder()
    {
        var resourceBuilder = ResourceBuilder
            .CreateEmpty() // Don't use CreateDefault because it puts service name unknown by default.
            .AddEnvironmentVariableDetector()
            .AddTelemetrySdk()
            .AddAttributes(new KeyValuePair<string, object>[] { new(Constants.Tracer.AutoInstrumentationVersionName, Constants.Tracer.Version) });

        var resource = resourceBuilder.Build();
        if (!resource.Attributes.Any(kvp => kvp.Key == "service.name"))
        {
            // "service.name" was not configured yet use the fallback.
            resourceBuilder.AddAttributes(new KeyValuePair<string, object>[] { new("service.name", FallbackServiceName()) });
        }

        // Moving the plugin code here and deleting the Configure method.
        var pluginManager = Instrumentation.PluginManager;
        if (pluginManager != null)
        {
            resourceBuilder.InvokePlugins(pluginManager);
        }

        return resourceBuilder;
    }
```

Reply:
@pjanotti, yes, 2 different approaches. I would still like to compare reflection case to the 2 times `build()`. 
2 times build sounds a bit more terrifying than reflection (in startup performance wise). There is a lot of stuff you need to do twice.